hydra-task/src/main/java/com/addthis/hydra/task/source/DataSourceQuery.java

```diff
@@ -0,0 +1,279 @@
+            Bundle next = pollAndCloseOnInterrupt(pollInterval, TimeUnit.MILLISECONDS);
+            if (next != null) {
+                return next;
+            }
+            if (closeFuture.isDone()) {
+                closeFuture.join();
+                return null;
+            }
+            if (pollCountdown > 0) {
+                log.info("next polled null, retrying {} more times. shuttingDown={}", countdown, shuttingDown.get());
+            }
+            log.info("null poll ");
+        }
+        if (countdown < 0) {
+            log.info("exit with no data during poll countdown");
+        }
+        return null;
+    }
+
+    private Bundle pollAndCloseOnInterrupt(long pollFor, TimeUnit unit) {
+        boolean interrupted = false;
+        try {
+            long remainingNanos = unit.toNanos(pollFor);
+            long end = System.nanoTime() + remainingNanos;
+            while (true) {
+                try {
+                    return queue.poll(remainingNanos, TimeUnit.NANOSECONDS);
+                } catch (InterruptedException e) {
+                    interrupted = true;
+                    log.info("interrupted while polling for bundles; closing source then resuming poll");
+                    close();
+                    remainingNanos = end - System.nanoTime();
+                }
+            }
+        } finally {
+            if (interrupted) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    @Override public Bundle peek() throws DataChannelError {
+        return queue.peek();
+    }
+
+    @Override public void close() {
+        if (shuttingDown.compareAndSet(false, true)) {
+            try {
+                if (queryHandle != null && queryHandle.isAlive()) {
+                    queryHandle.cancel("closing query input source");
+                }
+            } catch (Exception ex) {
+                firstError.compareAndSet(null, ex);
+                log.error("Error attempting to cancel query:", ex);
+            }
+            try {
+                if (tempDir != null) {
+                    Files.deleteDir(tempDir);
+                }
+            } catch (Exception ex) {
+                firstError.compareAndSet(null, ex);
+                log.error("Error attempting to delete temporary directory:", ex);
+            }
+            Throwable error = firstError.get();
+            if (error != null) {
+                closeFuture.completeExceptionally(error);
+                throw Throwables.propagate(error);
+            } else {
+                closeFuture.complete(null);
+            }
+        } else {
+            closeFuture.join();
+        }
+    }
+
+    @Override public void send(Bundle input) {
+        try {
+            Bundle output = input;
+            if (fields != null) {
+                output = output.createBundle();
+                for (int i = 0; i < fields.length; i++) {
```

Comment:
I think iterating over sequential index fields is the same as doing `for (BundleField field : bundle.getFormat())`